libbroker/broker/format/bin.hh

```diff
@@ -459,6 +512,399 @@ private:
   OutIter out_;
 };
 
+/// Decodes a broker::data or broker::variant from a binary representation. The
+/// decoding functions will generate a series of events for the handler, similar
+/// to a SAX parser.
+template <class Handler>
+std::pair<bool, const std::byte*>
+decode(const std::byte* pos, const std::byte* end, Handler& handler) {
+  if (pos == end)
+    return {false, end};
+  switch (static_cast<variant_tag>(*pos++)) {
+    case variant_tag::none:
+      handler.value(nil);
+      return {true, pos};
+    case variant_tag::boolean:
+      if (pos == end)
+        return {false, end};
+      handler.value(*pos++ != std::byte{0});
+      return {true, pos};
+    case variant_tag::count: {
+      auto tmp = uint64_t{0};
+      if (!read(pos, end, tmp))
+        return {false, pos};
+      handler.value(broker::count{tmp});
+      return {true, pos};
+    }
+    case variant_tag::integer: {
+      auto tmp = uint64_t{0};
+      if (!read(pos, end, tmp))
+        return {false, pos};
+      handler.value(static_cast<broker::integer>(tmp));
+      return {true, pos};
+    }
+    case variant_tag::real: {
+      auto tmp = 0.0;
+      if (!read(pos, end, tmp))
+        return {false, pos};
+      handler.value(tmp);
+      return {true, pos};
+    }
+    case variant_tag::string: {
+      auto size = size_t{0};
+      if (!read_varbyte(pos, end, size))
+        return {false, pos};
+      if (end - pos < static_cast<ptrdiff_t>(size))
+        return {false, pos};
+      auto str = reinterpret_cast<const char*>(pos);
+      pos += size;
+      handler.value(std::string_view{str, size});
+      return {true, pos};
+    }
+    case variant_tag::address: {
+      if (end - pos < static_cast<ptrdiff_t>(address::num_bytes))
+        return {false, end};
+      address tmp;
+      memcpy(tmp.bytes().data(), pos, address::num_bytes);
+      pos += address::num_bytes;
+      handler.value(tmp);
+      return {true, pos};
+    }
+    case variant_tag::subnet: {
+      static constexpr size_t subnet_len = address::num_bytes + 1;
+      if (end - pos < static_cast<ptrdiff_t>(subnet_len))
+        return {false, end};
+      address tmp;
+      memcpy(tmp.bytes().data(), pos, address::num_bytes);
+      pos += address::num_bytes;
+      auto length = uint8_t{0};
+      if (!read(pos, end, length))
+        return {false, pos};
+      handler.value(subnet::unchecked(tmp, length));
+      return {true, pos};
+    }
+    case variant_tag::port: {
+      if (end - pos < 3)
+        return {false, end};
+      auto num = uint16_t{0};
+      if (!read(pos, end, num))
+        return {false, pos};
+      auto proto = uint8_t{0};
+      if (!read(pos, end, proto))
+        return {false, pos};
+      if (proto > 3) // 3 is the highest protocol number we support (ICMP).
```

Comment:
Total nit, but if we know this is about ICMP we could just say `broker::port::protocol::icmp`.